services/api/service.go

```diff
@@ -381,6 +398,80 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(opts blockSimOptions) error {
+	t := time.Now()
+	simErr := api.blockSimRateLimiter.send(opts.ctx, opts.req, opts.highPrio)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.currentCounter(),
+	})
+	if simErr != nil {
+		log.WithError(simErr).Error("block validation failed")
+	} else {
+		log.Info("block validation successful")
+	}
+	return simErr
+}
+
+// Demotes a set of builders who have matching collateral IDs. This function
+// continues to try demoting even if an error is encountered.
+func (api *RelayAPI) demoteBuildersCollateralID(builderPubkey string) {
+	collateralPubkeys := []string{builderPubkey}
+	// Fetch builder collateral_id.
+	builder, err := api.db.GetBlockBuilderByPubkey(builderPubkey)
+	if err != nil {
+		api.log.WithError(err).Error("unable to get builder from database")
+	}
+	collateralID := builder.CollateralID
+
+	// Fetch additional builder pubkeys using the collateral_id.
+	otherBuilders, err := api.db.GetBlockBuildersFromCollateralID(collateralID)
+	if err != nil {
+		api.log.WithError(err).Error("unable to get extra builders from collateral id")
+	}
+	for _, b := range otherBuilders {
+		collateralPubkeys = append(collateralPubkeys, b.BuilderPubkey)
+	}
+
+	// Demote all the pubkeys in both redis and the db.
+	for _, pk := range collateralPubkeys {
+		err = api.redis.SetBuilderStatus(pk, common.LowPrio)
+		if err != nil {
+			api.log.WithError(err).Error("could not set builder status in redis")
+		}
+
+		err = api.db.SetBlockBuilderStatus(pk, common.LowPrio)
+		if err != nil {
+			api.log.WithError(err).Error("could not set builder status in database")
+		}
+	}
+}
+
+// startOptimisticBlockProcessor keeps listening on the channel and validating incoming blocks asynchronously.
+func (api *RelayAPI) startOptimisticBlockProcessor() {
+	for opts := range api.optimisticBlockC {
+		err := api.simulateBlock(opts)
+		if err != nil {
+			api.log.WithError(err).Error("block simulation failed")
+			builderPubkey := opts.req.Message.BuilderPubkey.String()
+			// Validation failed, demote all builders with the collateral id.
+			api.demoteBuildersCollateralID(builderPubkey)
+
+			bidTrace := &common.BidTraceV2{
+				BidTrace: *opts.req.Message,
+			}
+			// Upsert into the builder demotion table but without the
+			// blinded block or the validator registration, becuase we don't
+			// know if this bid will be accepted.
+			err = api.db.UpsertBuilderDemotion(bidTrace, nil, nil)
```

Comment:
I think it makes sense to have the other pieces of the builder submit block request: https://github.com/michaelneuder/mev-boost-relay/pull/1/commits/6692741804501afd1b2fb522bb0ca9ce1e88c08c implements this. basically we can save everything that the builder sent us, including their signature, the root hash, and the set of transactions that they want to include in the future block. I think this should be all the data we need to justify the demotion! IIUC we wouldn't need the validator registration, because this demotion has nothing to do with the validator. but correct me if i am wrong!

Reply:
> IIUC we wouldn't need the validator registration, because this demotion has nothing to do with the validator

My understanding is that some types of demotions do depend on the validator registration. Specifically, the validator registration specifies: 1) the requested fee recipient, and 2) the requested gas limit. A builder that does not respect the fee recipient or the gas limit should be demoted.

Reply:
that logic is already present in `handleSubmitNewBlock()`, and just results in an HTTP error being returned. https://github.com/flashbots/mev-boost-relay/blob/438c93935224d42e3954b42baea852ec40589d2b/services/api/service.go#L980-L991

IMO we should leave those prechecks alone and not necessarily demote someone whenever there is something wrong with there submission. It seems like demotion should only occur when the block simulation itself fails. WDYT?

Reply:
You're right—not demoting on prechecks makes a lot of sense! One reason is that this will minimise the number of false-positive demotions and therefore make the optimistic relay easier to maintain and operate.

Reply:
great :-)